cwl/src/main/scala/cwl/CwlType.scala

```diff
@@ -107,18 +110,44 @@ object File {
       ""
     }
   }
+  
+  private def recursivelyBuildDirectory(directory: String, ioFunctions: IoFunctionSet): WomMaybeListedDirectory = {
+    val listing: Iterator[WomFile] = sync(ioFunctions.listDirectory(directory)).map({
+      case entry if sync(ioFunctions.isDirectory(entry)) => recursivelyBuildDirectory(entry, ioFunctions)
+      case entry => WomMaybePopulatedFile(entry)
+    })
+
+    WomMaybeListedDirectory(Option(directory), Option(listing.toList))
+  }
 
   def secondaryStringFile(primaryWomFile: WomFile,
                           stringWomFileType: WomFileType,
-                          secondaryValue: String): ErrorOr[WomFile] = {
-    validate(WomFile(stringWomFileType, File.relativeFileName(primaryWomFile.value, secondaryValue)))
+                          secondaryValue: String,
+                          ioFunctions: IoFunctionSet): ErrorOr[WomFile] = {
+    val secondaryRelativeFileName = File.relativeFileName(primaryWomFile.value, secondaryValue)
+
+    // If the primary file is an absolute path, and the secondary is not, make the secondary file an absolute path and a sibling of the primary
+    // http://www.commonwl.org/v1.0/CommandLineTool.html#CommandInputParameter
+    val filePath = ioFunctions.pathFunctions.absoluteSibling(primaryWomFile.value, secondaryRelativeFileName)
+
+    // If the secondary file is in fact a directory, look into it and build its listing
+    val file = if (sync(ioFunctions.isDirectory(filePath))) {
+      recursivelyBuildDirectory(filePath, ioFunctions)
+    } else {
+      WomFile(stringWomFileType, filePath)
+    }
+    validate(file)
   }
 
   def secondaryExpressionFiles(primaryWomFile: WomFile,
                                stringWomFileType: WomFileType,
                                expression: Expression,
                                parameterContext: ParameterContext,
-                               expressionLib: ExpressionLib): ErrorOr[List[WomFile]] = {
+                               expressionLib: ExpressionLib,
+                               ioFunctions: IoFunctionSet): ErrorOr[List[WomFile]] = {
+    
+    def relativize(path: String) = ioFunctions.pathFunctions.absoluteSibling(primaryWomFile.value, path)
```

Comment:
`relativize` returns the result of `absoluteSibling`?

Reply:
Yeah the naming is confusing -_-
What it's doing is taking a "relative" path A and turning it into an absolute path B. It's doing that by saying, A is actually relative to absolute path C, so the absolute version of A is C/A.
Not sure how to describe that in terms of function names 😄